File: frontend/src/types/quiz.ts
/**
 * TypeScript type definitions for quiz-related data structures
 */

export interface QuestionOption {
  id: number;
  option_text: string;
  option_letter: string;
  is_correct: boolean;
}

export interface Question {
  id: number;
  question_text: string;
  question_order: number;
  options: QuestionOption[];
}

export interface Quiz {
  id: number;
  topic: string;
  created_at: string;
  questions: Question[];
}

export interface UserAnswer {
  question_id: number;
  selected_option_id: number;
}

export interface QuizSubmission {
  quiz_id: number;
  answers: UserAnswer[];
}

export interface QuizResult {
  score: number;
  total_questions: number;
  percentage: number;
  correct_answers: {
    question_id: number;
    question_text: string;
    correct_option: string;
    correct_text: string;
    user_selected: string;
    user_selected_text: string;
    is_correct: boolean;
  }[];
}

export interface QuizAttempt {
  id: number;
  quiz_id: number;
  attempted_at: string;
  score: number;
  total_questions: number;
}

// API response types
export interface ApiResponse<T> {
  data?: T;
  error?: string;
  message?: string;
}

// Component state types
export interface QuizState {
  currentQuestionIndex: number;
  answers: UserAnswer[];
  timeStarted?: Date;
  isSubmitting: boolean;
}

export interface AppState {
  currentQuiz: Quiz | null;
  isLoading: boolean;
  error: string | null;
  quizResult: QuizResult | null;
}

File: backend/app/models/quiz.py
"""
Database models for quiz-related entities.
"""

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from ..database import Base


class Quiz(Base):
    """Quiz model representing a generated quiz."""

    __tablename__ = "quizzes"

    id = Column(Integer, primary_key=True, index=True)
    topic = Column(String(100), nullable=False, index=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    # Relationships
    questions = relationship(
        "Question", back_populates="quiz", cascade="all, delete-orphan"
    )
    attempts = relationship(
        "QuizAttempt", back_populates="quiz", cascade="all, delete-orphan"
    )


class Question(Base):
    """Question model representing a quiz question."""

    __tablename__ = "questions"

    id = Column(Integer, primary_key=True, index=True)
    quiz_id = Column(Integer, ForeignKey("quizzes.id"), nullable=False)
    question_text = Column(Text, nullable=False)
    question_order = Column(Integer, nullable=False)

    # Relationships
    quiz = relationship("Quiz", back_populates="questions")
    options = relationship(
        "QuestionOption", back_populates="question", cascade="all, delete-orphan"
    )


class QuestionOption(Base):
    """Question option model representing multiple choice options."""

    __tablename__ = "question_options"

    id = Column(Integer, primary_key=True, index=True)
    question_id = Column(Integer, ForeignKey("questions.id"), nullable=False)
    option_text = Column(Text, nullable=False)
    option_letter = Column(String(1), nullable=False)  # A, B, C, D
    is_correct = Column(Boolean, default=False, nullable=False)

    # Relationships
    question = relationship("Question", back_populates="options")


class QuizAttempt(Base):
    """Quiz attempt model representing a user's attempt at a quiz."""

    __tablename__ = "quiz_attempts"

    id = Column(Integer, primary_key=True, index=True)
    quiz_id = Column(Integer, ForeignKey("quizzes.id"), nullable=False)
    attempted_at = Column(DateTime(timezone=True), server_default=func.now())
    score = Column(Integer, nullable=False)
    total_questions = Column(Integer, nullable=False)

    # Relationships
    quiz = relationship("Quiz", back_populates="attempts")
    responses = relationship(
        "UserResponse", back_populates="attempt", cascade="all, delete-orphan"
    )


class UserResponse(Base):
    """User response model representing a user's answer to a question."""

    __tablename__ = "user_responses"

    id = Column(Integer, primary_key=True, index=True)
    attempt_id = Column(Integer, ForeignKey("quiz_attempts.id"), nullable=False)
    question_id = Column(Integer, ForeignKey("questions.id"), nullable=False)
    selected_option_id = Column(
        Integer, ForeignKey("question_options.id"), nullable=False
    )
    is_correct = Column(Boolean, nullable=False)

    # Relationships
    attempt = relationship("QuizAttempt", back_populates="responses")
    question = relationship("Question")
    selected_option = relationship("QuestionOption")
